internal/resolution/dependency_subgraph.go

```diff
@@ -0,0 +1,244 @@
+	// Find the parent nodes of each node in graph, for easier traversal.
+	// These slices are shared between the returned subgraphs.
+	parentEdges := make(map[resolve.NodeID][]resolve.Edge)
+	for _, e := range g.Edges {
+		// Check for a self-dependency, just in case.
+		if e.From == e.To {
+			continue
+		}
+		parentEdges[e.To] = append(parentEdges[e.To], e)
+	}
+
+	// For each node, compute the subgraph.
+	subGraphs := make([]*DependencySubgraph, 0, len(nodes))
+	for _, nodeID := range nodes {
+		// Starting at the node of interest, visit all unvisited parents,
+		// adding the corresponding edges to the GraphNodes.
+		gNodes := make(map[resolve.NodeID]GraphNode)
+		seen := make(map[resolve.NodeID]struct{})
+		seen[nodeID] = struct{}{}
+		toProcess := []resolve.NodeID{nodeID}
+		currDistance := 0 // The current distance from end dependency.
+		for len(toProcess) > 0 {
+			// Track the next set of nodes to process, which will be +1 Distance away from end.
+			var next []resolve.NodeID
+			for _, node := range toProcess {
+				// Construct the GraphNode
+				parents := parentEdges[node]
+				gNode := gNodes[node] // Grab the existing GraphNode, which will have some Children populated.
+				gNode.Version = g.Nodes[node].Version
+				gNode.Distance = currDistance
+				gNode.Parents = parents
+				gNodes[node] = gNode
+				// Populate parent's children and add to next set.
+				for _, edge := range parents {
+					nID := edge.From
+					pNode := gNodes[nID]
+					pNode.Children = append(pNode.Children, edge)
+					gNodes[nID] = pNode
+					if _, ok := seen[nID]; !ok {
+						seen[nID] = struct{}{}
+						next = append(next, nID)
+					}
+				}
+			}
+			toProcess = next
+			currDistance++
+		}
+
+		subGraphs = append(subGraphs, &DependencySubgraph{
+			Dependency: nodeID,
+			Nodes:      gNodes,
+		})
+	}
+
+	return subGraphs
+}
+
+// IsDevOnly checks if this DependencySubgraph solely contains dev (or test) dependencies.
+// If groups is nil, checks the dep.Type of the direct graph edges for the Dev Attr (for in-place).
+// Otherwise, uses the groups of the direct dependencies to determine if a non-dev path exists (for relax/override).
+func (ds *DependencySubgraph) IsDevOnly(groups map[manifest.RequirementKey][]string) bool {
+	if groups != nil {
+		// Check if any of the direct dependencies are not in the dev group.
+		return !slices.ContainsFunc(ds.Nodes[0].Children, func(e resolve.Edge) bool {
+			req := resolve.RequirementVersion{
+				VersionKey: ds.Nodes[e.To].Version,
+				Type:       e.Type.Clone(),
+			}
+			ecosystem, ok := util.OSVEcosystem[req.System]
+			if !ok {
+				return true
+			}
+
+			return !lockfile.Ecosystem(ecosystem).IsDevGroup(groups[manifest.MakeRequirementKey(req)])
+		})
+	}
+
+	// groups == nil
+	// Check if any of the direct dependencies do not have the Dev attr.
+	for _, e := range ds.Nodes[0].Children {
+		if e.Type.HasAttr(dep.Dev) {
+			continue
+		}
+		if e.To == ds.Dependency {
```

Comment:
can you explain what this case is in a comment also? why does this imply there is a non dev usage?

Reply:
Added a comment and put this check after the workspace Children check to make more clear it's related.